### PR TITLE
refactor: select encoder

### PR DESCRIPTION
**Pull request by karasusan**
### Changed
- Added `WebRTC.Initialize()` method argument `encoderType`
- Added tests for Software/Hardware encoder using parameterized tests methods

## ~~TODO~~
- ~~Mac OS unit tests on Yamato are failed~~
    - ~~Test com.unity.webrtc with mono 2019.3 on macos~~
    - ~~Test com.unity.webrtc with il2cpp 2019.3 on macos~~

### CODE REVIEW

File: Tests/Runtime/MediaStreamTest.cs [hunk 1]
```diff
@@ -1,101 +1,201 @@
...
+            foreach (var track in videoStream.GetTracks())
+            {
+                pc1Senders.Add(peer1.AddTrack(track));
+            }
 
-        yield return videoStream.FinalizeEncoder();
-        videoStream.Dispose();
+            RTCOfferOptions options1 = default;
+            RTCAnswerOptions options2 = default;
+            var op1 = peer1.CreateOffer(ref options1);
```
Comment (sindharta) [inline Tests/Runtime/MediaStreamTest.cs]: Can we make a function to share some code with MediaStreamTest_AddAndRemoveAudioTrack() ?

File: Tests/Runtime/NativeAPITest.cs [hunk 2]
```diff
@@ -1,112 +1,168 @@
...
-    public void CreateAndDestroyContext()
-    {
-        var context = NativeMethods.ContextCreate(0);
-        NativeMethods.ContextDestroy(0);
-    }
+        [SetUp]
+        public void Init()
+        {
+            NativeMethods.RegisterDebugLog(DebugLog);
+            encoderType = EncoderType.Software;
```
Comment (sindharta) [inline Tests/Runtime/NativeAPITest.cs]: should be Hardware ? The name of the class is NativeAPITestWithHardwareEncoder

File: Tests/Runtime/NativeAPITest.cs [hunk 3]
```diff
@@ -1,112 +1,168 @@
...
+            Assert.AreNotEqual(callback, IntPtr.Zero);
+            NativeMethods.ContextDestroy(0);
+            NativeMethods.GetRenderEventFunc(IntPtr.Zero);
+        }
+
+        [UnityTest]
+        public IEnumerator CallVideoEncoderMethods()
+        {
+            var context = NativeMethods.ContextCreate(0, encoderType);
+            const int width = 1280;
```
Comment (sindharta) [inline Tests/Runtime/NativeAPITest.cs]: Maybe we can make a function to share some code with CreateAndDeleteVideoStream() ?

File: Tests/Runtime/WaitUntilWithTimeout.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+                return false;
+            }
+
+            return !(Time.realtimeSinceStartup >= timeoutTime);
+        }
+    }
+
+    public WaitUntilWithTimeout(System.Func<bool> predicate, int timeout)
+    {
+        this.timeoutTime = Time.realtimeSinceStartup + timeout * 0.001f;
```
[the remark this answers is not in the document]
Comment (karasusan) [reply Tests/Runtime/WaitUntilWithTimeout.cs]: The unit "timeout" is msec.

File: Tests/Runtime/NativeAPITest.cs (hunk 3 above)
Comment (karasusan) [reply Tests/Runtime/NativeAPITest.cs]: I created `CreateRenderTexture`.

File: Tests/Runtime/NativeAPITest.cs (hunk 2 above)
Comment (karasusan) [reply Tests/Runtime/NativeAPITest.cs]: Fixed.

File: Tests/Runtime/MediaStreamTest.cs [hunk 5]
```diff
@@ -1,101 +1,201 @@
...
-        });
-        var videoStream = cam.CaptureStream(1280, 720);
-        yield return new WaitForEndOfFrame();
+            WebRTC.Finalize();
+        }
 
-        foreach (var track in videoStream.GetTracks())
+        [UnityTest]
+        [Timeout(5000)]
+        [UnityPlatform(exclude = new[] {RuntimePlatform.OSXEditor})] // TODO:: fix Crush bug on Yamato
```
[the remark this answers is not in the document]
Comment (karasusan) [reply Tests/Runtime/MediaStreamTest.cs]: https://github.com/Unity-Technologies/com.unity.webrtc/issues/59

File: Tests/Runtime/MediaStreamTest.cs [hunk 6]
```diff
@@ -1,101 +1,201 @@
...
+            audioStream.Dispose();
         }
-        pc1Senders.Clear();
-        GameObject.DestroyImmediate(camObj);
 
-        peer1.Close();
-        peer2.Close();
+        [UnityTest]
+        [Timeout(5000)]
+        [UnityPlatform(exclude = new[] {RuntimePlatform.OSXEditor})] // TODO:: fix Crush bug on Yamato
```
[the remark this answers is not in the document]
Comment (karasusan) [reply Tests/Runtime/MediaStreamTest.cs]: https://github.com/Unity-Technologies/com.unity.webrtc/issues/59

**Comment by sindharta** [comment]
↑
[10/53] TestFinish
	name: Unity.WebRTC.RuntimeTest.MediaStreamTest.MediaStreamTest_AddAndRemoveVideoStream
	result: FAILED
	message: Unhandled log message: '[Exception] NullReferenceException: Object reference not set to an instance of an object'. Use UnityEngine.TestTools.LogAssert.Expect
	duration: 1.899525 seconds

**Comment by sindharta** [comment]
Current error:
[03:39:27.077 INF] [10/53] TestFinish
	name: Unity.WebRTC.RuntimeTest.MediaStreamTest.MediaStreamTest_AddAndRemoveVideoStream
	result: FAILED
	message: Test exceeded Timeout value of 5000ms
	duration: 5.01299 seconds

**Comment by sindharta** [comment]
LGTM

**Pull request merged**